### PR TITLE
fix(ext/fs): retry if file already exists in makeTempFile

### LINKED ISSUE #17781: Deno.makeTemp* do not appear to be checking for file existence
I've run this on Deno 1.28.2, but it doesn't look like it's a recent regression from a cursory look at`git log` and `git blame`. Simple repro:

```
$ mkdir out
$ cat repro.ts
while (true) { console.log(Deno.makeTempFileSync({dir: "out"})); }
$ deno run --allow-all repro.ts # this will generate ~100k files, beware!
...
out/017f5cd8
out/b9d7898a
out/c2c8ca65
out/ea05084a
out/1b342639
out/671325db
out/9703a9d5
out/28d6be21
error: Uncaught AlreadyExists: File exists (os error 17)
while (true) { console.log(Deno.makeTempFileSync({dir: "out"})); }
                                ^
    at Object.makeTempFileSync (deno:runtime/js/30_fs.js:88:16)
    at file:///Users/cscheid/Desktop/daily-log/2023/02/14/deno-temp-file-stress/main.ts:1:33
$ cd out
$ ls | wc
   77678   77678  699102
```

So, although `make_temp` [has code that attempts to detect AlreadyExists](https://github.com/denoland/deno/blame/main/runtime/ops/fs.rs#L1778-L1782), that code does not appear to be executing.

We have seen this actually happen in the wild at least once. It's not as rare as one would naively expect because of the birthday paradox. We saw it when generating on the order of a few thousand files in the same directory. That was particularly unlucky but the point still stands.

### PR DESCRIPTION
**Pull request by 0f-0b**
Closes #17781.

### CODE REVIEW

File: runtime/ops/fs.rs [hunk 1]
```diff
@@ -1772,8 +1772,7 @@ fn make_temp(
         use std::os::unix::fs::OpenOptionsExt;
         open_options.mode(0o600);
       }
-      open_options.open(buf.as_path())?;
-      Ok(())
+      open_options.open(buf.as_path()).map(drop)
```
Comment (dsherret) [inline runtime/ops/fs.rs:1775]: Would this cause an infinite loop if this failed every time?

**Review by bartlomieju — COMMENTED** [review]
@0f-0b I think we should maybe try for maximum of N times (maybe 10 times?) to create a temp directory and bail out with an error if that doesn't work. Otherwise we're vulnerable for denial of service if someone stuff temp directory with a huge number of phony directories.

**Comment by bartlomieju** [comment]
@dsherret I'm in favor of landing this one. Any objections?

File: ext/fs/lib.rs [hunk 2]
```diff
@@ -1878,15 +1879,18 @@ fn make_temp(
...
     };
     match r {
       Err(ref e) if e.kind() == std::io::ErrorKind::AlreadyExists => continue,
       Ok(_) => return Ok(buf),
       Err(e) => return Err(e),
     }
   }
+  Err(io::Error::new(
+    io::ErrorKind::AlreadyExists,
+    "too many temp files exist",
```
[the remark this answers is not in the document]
Comment (0f-0b) [reply ext/fs/lib.rs:1892]: I think the error message is OK for that it directly points out the problem which the user should resolve, and that an error message mentioning implementation details such as collisions and retries would be less helpful. `tempfile` uses [a similar message](https://docs.rs/tempfile/3.4.0/src/tempfile/util.rs.html#47).

[the remark this answers is not in the document]
Comment (dsherret) [reply ext/fs/lib.rs:1892]: > it directly points out the problem which the user should resolve

Isn't it possible for this to fail one time and then pass on the next time though?

[the remark this answers is not in the document]
Comment (dsherret) [reply ext/fs/lib.rs:1892]: I guess given `tempfile` does this that's good precedence and this probably likely will never happen for anyone.

**Pull request merged**